### ISSUE TITLE
Port Recalbox to x86 (Recalbox live usb)

### ISSUE THREAD

**Issue by nadenislamarre** [recalbox/recalbox-os#999]
Current status :
- works with some manual steps on my computer : boot from the sd/start es over X/play snes retroarch
- need still some work to make the boot process automatic
- need some work to add other emulators
- kodi is working
- 64 bits would be nice to add gamecube/wii and why not ps2
- an idea is to make an entry on any recalbox version to choose a usb key, download the x86 version and install it on the usb key (making it really easy to install)
- uefi is not wanted for the moment. buildroot allows with or without uefi, so, 2 images would be required.
- only the linux supported hardware from scratched is supported (mainly intel/ati/nvidia gpu)

**Comment by nadenislamarre** [comment]
A first video of the work in progress : http://recalbox.remix.free.fr/videos/x86.avi

**Comment by sebastianheyn** [comment]
What are you basing it on? Scratch build? Do you have a build script?

**Comment by nadenislamarre** [comment]
Recalbox is based on buildroot.
The build process is exactly the same than for rpi and odroids.
You can compile in 6 commands :
See : https://github.com/recalbox/recalbox-os/wiki/Recalbox-development---basics-%28EN%29

The port on x86 is just some configuration of the boot process (grub instead of uboot for example), some compilation options changed, some driver configuration changed (add the pci cards for example), some adaptation of some part of the boot process (on rpi and xu4, we consider that the boot comes from a sd card), ...

See :
https://github.com/recalbox/recalbox-os/wiki/recalboxOS-detailled-%28EN%29

**Comment by sylvain-perez** [comment]
Very nice!

**Comment by nadenislamarre** [comment]
A lot of things have been done these days.
90% of the job is done.

But to resume : what a hell to boot on a live usb on x86 ! I coulnd't have guessed that it was so complicated.

What works :
- It works very nicely on my computer (and no code is specific to my computer).
- sound, wifi, gpu, pads : out of the box
- kodi works perfectly
- emulationstation works
- only 1 required emulator doesn't compile (Scumm)
- all tested games works, psx is slow on my notebook
- tested emulators : snes, nes, gameboy, mastersystem, megadrive, psx, neogeo
- tested computer : https://fr.wikipedia.org/wiki/Aspire_One A110L
  * Intel Atom 800 MHz
  * Chipset Mobile Intel 945GSE Express

I can start to list what is missing to finish the job :

1) Mandatory for a 1st test image :
- boot: initrd to not have to choose manually the /dev/sdx to boot on
- linux: adjust linux drivers configuration (config.in)
- share: make recalbox forget /dev/mmcblk0 / internal_device and base the value on rdev

2) Mandatory for a first usable image :
- es : keyboard works but not in textbox
- hardware : wiimote patch
- emulator : Scumm (compilation fails) (not analysed)
- emulator : reicast libzip path to adjust
- emulator : nintendo 64 compile but doesn't start (not analysed)
- emulator : zoom is not correct (too zoomed (sega) or not enought zoomed (gameboy))
- acpi : my power button doesnt ask the recalbox to shut down

3) Not strictly required :
- emulator : adjust resolution for performances (lower resolution / list of available resolutions)
- x86_64 (will be in an other issue)
- dolphin for x86_64 (will be in an other issue)

Boot process on x86 :
1) the bios boots on the live usb key
2) grub takes the power and loads the kernel
3) the kernel has not enough information to know the device which was used to boot (and the 2nd grub disk, it not necessarly the 2nd kernel disk, nor the 2nd bios disk)
4) so, use an initramfs ? => no because for buildroot integration, an external initrd is easier
5) the kernel boots on the initrd => not working for the moment
6) the initrd loads usb/sd drivers ? => for the moment, i did n't compile them as modules, but directly into the kernel
7) the initrd looks for the RECALBOX partition by label to boot on it => so, if there are several one, take the last one to priorize usb live key to internal hard drive having recalbox installed
8) once linux is booted, look for the boot device, which is root-1 and the internal share device which is root+1 by convention (/dev/sda1 sda2 sda3 for example)
9) how to find the root device ? it looks easy, isn't it ? but linux can boot on an initramfs without mounting anything, or from a network share, so, there is not an easy information in /proc /sys : here comes rdev looking for /dev devices and / informations ;-)
10) and now, the recalbox can finally start !
11) hopefully, buildroot and linux rocks ! and all is almost already working.

Boot process on the rpi :
- the rpi boots, find the 1st fat partition
- it loads the kernel according to the configuration file
- it mounts the root device according to the configuration file (it's easier, there is only 1 possible device : the sd card ;-)
- the recalbox starts

on the xu4, it's almost like on the rpi, but a boot loaded (uboot) is at the begining of the sdcard to load the kernel, and a proprietary binary is at the beginning of the sdcard to load the boot loader.

**Comment by substring** [comment]
Nice job !

I wouldn't rush in x86_64. Some emulators seem not to work on a 64bits architecture. Even lakka didn't port to 64bits yet because of that. That's way too much work

How do reicast and ppsspp work ?
Concerning the resolution, as we've always relied on tvservice to change the screen resolution, you may face the same problem as on xu4 : can't change the resolution on the fly

**Comment by nadenislamarre** [comment]
reicast ans ppsspp should work. not tested.
at least they compiles (after some minor changes).

i thing that with x, i can detect and change the resolutions on the fly, but i've not checked.
I was to make my first image quickly so that you can help me to finnish the job ;-)
But hollidays are coming soon, so probably, it will be only for september.

**Comment by substring** [comment]
X ??? running ES on X ???

**Comment by nadenislamarre** [comment]
yes, running es on x.
and i'm asking myself why we don't use it in general for rpi/xu4.

**Comment by substring** [comment]
for performance and memory management ! you should really try to avoid running X, that adds too many differences with the "original spirit" of recalbox + increased support difficulties

**Comment by nadenislamarre** [comment]
i think that there is a misunderstanding.
the memory footprint is low. I just start xinit, not the tools around it.
X doesn't contains a window manager nor a desktop environment. It starts immediatly.
for performances, x seems to be an advantage (because framebuffer usecase is not games).
for maintaining, x seems to be an advantage, because it has a common api.
but i don't avoid to go to wayland which is the future.

On x86, after several reads and talks on irc channels, my conclusion is that the proper way is to go throw X.
The only other possibility (except the abandonned directfb) is drm which is even not supported by sdl2.

I found no advantage to avoid x. In fact, for the moment, i think it has only advantages on the framebuffer,
My leading rule is : if it contains a X, do it. It includes recalbox, xu4, x86, linux and xorg ;-)

**Comment by nadenislamarre** [comment]
note that adding X doesn't change a lot of things in recalbox :
- add some X checkbox in _defconfig of x86
- change in S31emulationstation : /usr/bin/emulationstation by xinit /usr/bin/emulationstation

but we'll see.
However, as told, i've found no other way in x86.

**Comment by nadenislamarre** [comment]
Some arguments can be found here for example :
http://stackoverflow.com/questions/3327767/why-is-directfb-not-more-widely-used-in-gnu-linux-are-there-crippling-limitatio

**Comment by nadenislamarre** [comment]
Please find a first commit here :
https://github.com/nadenislamarre/recalbox-buildroot/commit/994ab8c6467a0e11e0e603aafbc3c01744711bbe

you can compile and boot it, it will work.
i don't give the image because the only aim of this commit is to get first feedback on the code, and why not help on fixing issues.

@substring reicast and ppsspp doesn't work in fact. it compiled, but it doest compile anymore. i've not checked why ; some path issue about zlib and such things.

**Comment by substring** [comment]
Nice job pal :)

Quite some things, most of them are not important at all, just curiosity or cosmetics :
* you solved some odroid stuff in the PR, would have been better in a separate commit
* you should start using some one liner for things like on https://github.com/nadenislamarre/recalbox-buildroot/commit/994ab8c6467a0e11e0e603aafbc3c01744711bbe#diff-0be6d4235995552d62a6661743abaa0aR20 and drop that test command ;)
* why switch input to sdl2 for retroarch ? or is it just because the default retroarch.cfg file is set for sdl2 ?
* rdev works on pi and xu4 ?
* have you compiled for pi with this commit, to make sure there are no big regressions ?
* personnal opinion : one S31 for every platform would be easier to maintain, dunno what @digitalLumberjack thinks about that
* why so many entries for Grub ???
* buildroot for x86 uses a 4.5 kernel ?
* couldn't check yet, but it looks like the x86 defconfig is almost the same as pi and xu4 ?
* why allow the possibility to disable ES on x86 ?
* so far we are not using libretro-uea. If it compiles, it's a mistake, should change the defconfigs
* why use GLES when you can use real OpenGL on x86 ?


General thought : we should consider new mame cores for xu4 and x86 (2010 and 2014, see if they are worth)

**Comment by nadenislamarre** [comment]
thanks for this interesting feedback.

- you solved some odroid stuff in the PR, would have been better in a separate commit
arg, just a minor changed, effectively. i forgot it.

-    you should start using some one liner for things like on nadenislamarre/recalbox-
i don't like oneline when it's not basic. but don't hesitate to update ;-)

-    why switch input to sdl2 for retroarch ? or is it just because the default retroarch.cfg file is set for sdl2 ?
we should always use sdl2 while recalbox uses es which is based on sdl2.
if i don't force sdl2, it uses gl and not x. sdl2 is a way to let decide depending on what the system allow.

-    rdev works on pi and xu4 ?
yes. It's not specific.

-    have you compiled for pi with this commit, to make sure there are no big regressions ?
the test will be made later. But i test "quickly" the commands and scripts on pi2 to check, yes.

-    personnal opinion : one S31 for every platform would be easier to maintain
it's just a 5 lines file. But for the moment, i've no a good way to make only one nice file.

-    why so many entries for Grub ???
just because the initramfs is not written, so the job is not done automatically. To test, you need to select your usb key.

-    buildroot for x86 uses a 4.5 kernel ?
probably. I don't think i've choosed something.

-    couldn't check yet, but it looks like the x86 defconfig is almost the same as pi and xu4 ?
yes. almost the same as xu4. rpi have some rpi specific packages.

-    why allow the possibility to disable ES on x86 ?
it already exists on the current version. I've changed nothing.

-  so far we are not using libretro-uea. If it compiles, it's a mistake, should change the defconfigs
probably i've some thing i could disabled. i don't know what uea is for.

-    why use GLES when you can use real OpenGL on x86 ?
i've probably not disabled gles everywhere.

**Comment by nadenislamarre** [comment]
it works without any human intervention now. it is great.

done =>    boot: initrd to not have to choose manually the /dev/sdx to boot on
done =>    share: make recalbox forget /dev/mmcblk0 / internal_device and base the value on rdev
via : https://github.com/nadenislamarre/recalbox-buildroot/commit/33668e9f70ca9010cfe80df1201d77ac7ab59c49

next step :
to be done :    linux: adjust linux drivers configuration (config.in)

**Comment by nadenislamarre** [comment]
anybody wants to help me, for example, on the emulators part ?
i've finished the most complex parts.

so that i can work on x86_64, c2 and pine64 ?
Otherwise, i will finnish x86 before working on the rest.

Remaining work to do :
- linux: adjust linux drivers configuration (linux .config)
- x86_defconfig : remove gles. review the package list.
- es : keyboard works but not in textbox
- hardware : wiimote patch
- emulator : Scumm (compilation fails) (not analysed)
- emulator : reicast libzip path to adjust
- emulator : nintendo 64 compile but doesn't start (not analysed)
- emulator : zoom is not correct (too zoomed (sega) or not enought zoomed (gameboy))
- acpi : my power button doesnt ask the recalbox to shut down
- emulator : adjust resolution for performances (lower resolution / list of available resolutions)
- review the upgrade process (the progression bar via pygame doesn't start under x)

**Comment by nadenislamarre** [comment]
linux: adjust linux drivers configuration (linux .config) : done. Need to be tested.

i've taken the configuration file from ubuntu while they enable all the correct things.

I've just put the mass storage drivers into the kernel instead of modules to make things simpler (not have to integrate the modules into the initramfs).
I've kept the sd card readers driver as modules ; i don't know if some computer can boot from an sd card.
It makes 135 MB of drivers. Several thousands of supported hardware. Probably too much (like video recorder, mouses, gadgets, cryptology, ...) but i want to keep it easy to maintain. So for the moment, ubuntu config + mass storage not as a module.

I'm building a fresh image. If it works, i'll make it available before my holidays.
In parallel, i've started the x86_64 branch. It's building, but unfortunally, the night is not long enough to compile 2 recalbox (x86 + x86_64). However, all seems to compile (including enabled emulators).

@substring any information about non working emulator on x86_64 ?

**Comment by nadenislamarre** [comment]
@substring my chineese ps3 controller which works badly on xu4/rpi works nicely on the x86 with a recenter kernel and the same bluez version.

**Comment by substring** [comment]
With bluez4 ? Or bluez5 ? Shanwan or gasia ?

**Comment by nadenislamarre** [comment]
@substring all by default. I've changed nothing. the recalbox is just up.

**Comment by nadenislamarre** [comment]
Please find a first version for x86 : http://recalbox.remix.free.fr/4/x86/test/recalbox.img.gz
Remember, it's just for computers having a bios (so basically, not most of computer after 2010).
Remember, this is in alpha stage : most of things doesn't work.
The aim is to confirm it boots and that hardware works.

**Comment by Supernature2k** [comment]
Hi Susan33, here are the results of my test:

First of all I had to rename the file to recalbox.img to burn it with Win32DiskImager.

My config:

Core i5 4570 @3.20GHz on H87 Plus chipset
My UEFI can launch USB in UEFI or BIOS Mode :)
2 GPU:

Intel HD Graphics 4600 (i5 integrated)
GTX 750 Ti

When I run it with Intel HD: works smoothly, games launch, USB gamepad is OK
WHen I run with GTX, loading takes ages, the interface is laggy and ultra slow. Games launch but unplayable obviously.

**Comment by orion2036** [comment]
Hi Susan33 and thank you for your job on several versions of Recalbox.
I've just tested this x86 version with my following config:

Amd Athlon 62 X 2  Dual core 3800+
Radeon HD4200
Wired Bigben PS3PAD

Emulationstation was booting when the screen shut off and i had to push the on/off button of the pc to restart the screen, looks like Recalbox entered in standby mode right after Recalbox loading.
Emulationstation is very slow but working.

**Comment by substring** [comment]
P4 HT + 7600GT : I get the Recalbox at grub, then no display. Looks like it prepared the share, but i neverhad any display
i5 6500 + GTX 760 : I get the Recalbox at grub, scans all disks -> The background turns to green, and it's frozen. With the previous build I could reach ES

**Comment by Minimarko32** [comment]
RecalBox x86/amd64

Cool with Ubuntu core !

**Comment by nadenislamarre** [comment]
ctrl alt f3 to get messages
i guess ubuntu disable ati nvidia drivers to colpile externzlly i ll check
later

**Comment by nadenislamarre** [comment]
Hi,
i've created an alpha3.
I've no way to check if my changes solve the gpu performances issues.
So, please test.
Don't expect to play while since i disabled gles, retroarch won't ldd anymore. I've not fixed.

Aim of this build : get emulationstation starting with good performances.

http://recalbox.remix.free.fr/4/x86/alpha

**Comment by nadenislamarre** [comment]
@substring : nv xorg driver is now included.
as for other alpha, intel gpu works.

**Comment by nadenislamarre** [comment]
note that you can create a virtualbox image like this :
vboxmanage convertfromraw ./recalbox.img vmdkname.vmdk --format VMDK

**Comment by nadenislamarre** [comment]
just extract and copy http://recalbox.remix.free.fr/4/x86/alpha/EFI.tar.gz on the BOOT partition
(to get a EFI directory at the top of the boot partition),
and you'll be able to boot from an efi x86_64 computer ;-)
and you know what ? with my nvidia card, all rocks !
so, perhaps, it is just ati cards that are suffering.
More feedbacks are welcome.

**Comment by orion2036** [comment]
ES is still slow with the alpha 3 on my pc:
Amd Athlon 62 X 2 Dual core 3800+
Radeon HD4200
Wired Bigben PS3PAD

**Comment by nadenislamarre** [comment]
the x86_64 image alpha1 is done too. It works on my pc with a nvidia card.
i pushed a x86_64 branch for curious people.

https://github.com/nadenislamarre/recalbox-buildroot/commits/rb-4.1.X-nadenislamarre-x86_64
http://recalbox.remix.free.fr/4/x86_64/alpha/

**Comment by Blist** [comment]
Hi, 

How do you install it ?
I try to make a live usb with 

`dd if=recalbox.img of=myusb bs=1m` 

but impossible to boot on it. (I can boot with a ubuntu live cd, this is characteristics of this pc  http://support.hp.com/ch-fr/document/c01927359)
Is it possible to install it on a hard drive ? if yes how ?

Thanks a lot for your work ! it's look be awesome 👍

**Comment by iNgoes** [comment]
Just try for alpha 1 & 3.
For emulator, works great for alpha 1 but can't play anything with alpha 3. And for the gamelist, I just have a list, not with game description like your video above, how to fix these?

Anyway, you're the man, thanks to make this works for my old netbook :+1:

**Comment by nadenislamarre** [comment]
alpha 3 has games disabled.
i just wirk on the boot for the moment

**Comment by ST68k** [comment]
One more great job from nadenislamarre !
Alpha version works fine on my 2 laptops of 2011 :
- intel core i7 with nvidia graphics
- intel core i5 with intel graphics
Another great breakthrough for recalbox !
See you guyz !

**Comment by ST68k** [comment]
One more great job from nadenislamarre !
 Alpha version works fine on my 2 laptops of 2011 :
 •intel core i7 with nvidia graphics
 •intel core i5 with intel graphics

Another great breakthrough for recalbox !
See you guyz !

**Comment by nadenislamarre** [comment]
Alpha 4 : http://recalbox.remix.free.fr/4/x86/alpha

Now you can play ;-)
Most emulator (based on retroarch in fact) are working.

A rule : if it boots and works, your computer is supported ;-)

@substring can you retry but not by restarting your computer. By shutting it down and boot it. It seems that there is a difference in a disk detection for grub.

**Comment by Kivutar** [comment]
I tested the image and it works well in most cases.

A few remarks (maybe not x86 specific):

 * The volume up/down keys of my laptop keyboard where not recognized, and the background music volume pretty high. It took me the time to configure a controller and lower the volume.
 * I got a crash accessing system settings. I had to force reboot after that.
 * Switching the ES theme gave me a white screen
 * Boot logo was glitched
 * First boot took long time with a black screen (maybe 4 minutes?)
 * Tried to transfer some roms by directly mounting the USB drive on Linux, but the folders permissions didn't allow me to paste any file in there.

**Comment by nadenislamarre** [comment]
thanks for the feedback.
for the first boot, it depends on your usb stick. it is finishing to initialize the card (creating the 3rd partition, and initializing files)
for the transfert, sudo ;-) or via the network.
for the rest, i'm make some tests.

**Comment by nadenislamarre** [comment]
Ati cards should work better now, thanks orion 2036 ;-)

# alpha 5 :
- n64 is now working
- add several firmware (mainly for ati gpu cards but for wifi/bluetooth too)
- more chance to get a splash screen

download : http://recalbox.remix.free.fr/4/x86/alpha/
commits : https://github.com/nadenislamarre/recalbox-buildroot/commits/rb-4.1.X-nadenislamarre-x86

**Comment by NickDG** [comment]
Hi , 

First of all i just want to say thanks to nadenislamarre for all the hard and awesome work on this recalbox to x86

now the issue, im not sure if im the only one but im unable to open the following link 

http://recalbox.remix.free.fr

id like to download the latest version, but when i click on the link all i get is a blank page

**Comment by nadenislamarre** [comment]
the direct link is : http://recalbox.remix.free.fr/4/x86/alpha/alpha5/recalbox.img.gz
works for me.

**Comment by nadenislamarre** [comment]
@digitalLumberjack may i have an account on recalbox ? or may we add an on demand build on the server ?

**Comment by digitalLumberjack** [comment]
I can create an automated build. The account would not be usefull as the jenkins configuration is stored in the docker configuration (manual changes are not lasting)

As soon as you make a PR i'll make a new build :)

**Comment by digitalLumberjack** [comment]
@nadenislamarre huge job here pal, i'm impressed and pleased to test recalbox on my personal computer :)

**Comment by nadenislamarre** [comment]
# alpha 6 :
- ppsspp is now working
- fix acpi power button
- externalize efi

http://recalbox.remix.free.fr/4/x86/alpha/

**Comment by orion2036** [comment]
I think it should be interesting to make a CD or DVD version of Recalbox x86. Once the BIOS is configured to boot on CD/DVD drive first, newbies won't have to enter inside the BIOS anymore.

**Comment by nadenislamarre** [comment]
a cd/dvd is not very interesting.
you've to copy roms to be able to play and save them between 2 sessions.

**Comment by nadenislamarre** [comment]
# alpha 7 :
this version is particularly interesting because almost everything is working now.
It's really working well.

- add a window manager to fix several issues about windows/inputs (including retroarch window sizing, kodi inputs)
- fix the wiimote/wiimote classical controller and some other joysticks
- handle multimedia keys (volume up/down/mute)

http://recalbox.remix.free.fr/4/x86/alpha/

**Comment by orion2036** [comment]
Roms can be copied with CD-RW or DVD-RW

**Comment by nadenislamarre** [comment]
this is not the target. It must be user friendly.

**Comment by daniel3d** [comment]
great work nadenislamarre do you have a git repo if not do you think is good idea creating one this way we can help.

**Comment by nadenislamarre** [comment]
Something like https://github.com/recalbox/recalbox-buildroot ?
;-)

**Comment by daniel3d** [comment]
I was able to run it inside virtualbox :) but the ui is really slow is virtualbox issue im not sure how is going to run under hyperv

![capture](https://cloud.githubusercontent.com/assets/1344188/18590292/2fa1b24c-7c26-11e6-9525-16a7d6a642d5.JPG)

**Comment by nadenislamarre** [comment]
yes, it works inside virtualbox (slowly).
But to make it fast, it must include the virtualbox driver which is version
dependant,
so, in practice, don't expect to get it fast in virtualbox.

**Comment by nadenislamarre** [comment]
# alpha 8 :
- fix the keyboard
- dosbox (pc) is now working
- apple2 is now working
- reicast should be working (require opengl 3.1)

http://recalbox.remix.free.fr/4/x86/alpha/

dosbox works but require a good configuration that i forgot to save. I don't find to make it work again ;-(

**Comment by ron15** [comment]
This is just what im looking for. I have tried this with two systems. My AMD E-350 System did great in running this port the only problem is no audio both is ES/retroarch and kodi. But so far everything runs smooth. On my intel i5 4600 System the USB drive booted but it was stucked on the splash screen (the one with the small logo with black background).

**Comment by nadenislamarre** [comment]
For the sound card, i known that on some computer sounds doesn't work. I
will try to find the root cause, and see if i can fix it.
Can you however run /recalbox/script/recalbox-support.sh and send me the
resulting zip file please.

For the stucked splash, i've never seen it.
Can you press : ctrl+alt+f3 so see the last log lines ?
Can you press : ctrl+alt+f2 (or f1, not sure), to see if you can connect
with the root/recalboxroot login ?

**Comment by ron15** [comment]
Here's the zip file from the system without sound. 
[recalbox-support-20160919141743.tar.gz](https://github.com/recalbox/recalbox-os/files/479941/recalbox-support-20160919141743.tar.gz)

For the stuck splash, its still to follow cause that's my main pc. The one where I'm typing this reply.. =)

**Comment by ron15** [comment]
Unfortunately, It's really stuck in the splash screen. I seems that it hanged and pressing ctrl+alt+f1,f2,f3 doesn't do anything.

Would it be possible that it is because of my dedicated GPU r9 380?? Im too lazy to remove my GPU and test it using the integrated graphics. hahaha... XD

**Comment by nadenislamarre** [comment]
open the file on the key :
grub.cfg (there are 2 in fact, but 2 times the same, change both)
and change :

and remove the 3 parameters :
quiet loglevel=0 tty=3

to see on what it fails;

**Comment by ron15** [comment]
I have removed the parameters which you have said. Here's an actual screen shot of the last boot log. 
![ss01](https://cloud.githubusercontent.com/assets/22293314/18634166/78858338-7eb2-11e6-9d3e-f3991b5e7011.jpg)

**Comment by nadenislamarre** [comment]
can you try by adding the following parameters where you removed the others
:

noapic nolapic acpi=off

**Comment by nadenislamarre** [comment]
I got trouble with 2 computers having no sound.
The common point was the sound chipset.
The are 2 problems, but i didn't found a correct way to fix it.
So for the moment, it requires a manual action described below.

1) For an unknown reason, the driver is not loaded (while the pci values are referenced in the driver)
2) the default output are not the speakers

mount -o remount,rw /
echo "snd_hda_intel" > /etc/modules.conf
echo ""options snd-hda-intel enable=0,1" > /etc/modprobe.d/alsa-intel.conf
mount -o remount,ro /
shutdown -r now

**Comment by nadenislamarre** [comment]
@substring @ron15 
for crashing computers working on ubuntu, can you try the following :
copy /etc/modprobe.d from the ubuntu to /etc/modprobe.d/ on the recalbox.

It contains a list of drivers to blacklist ; while i've the same enabled drivers than on ubuntu, the crashing driver could be present in this list.

**Comment by nadenislamarre** [comment]
the remaining todo list for x86 :
- scumvm : for the moment i will not fix it while the same package has troubles too with xu4 and i've not managed to fix it easyly (fixing the error fall in next error, then in next error, ...)
- sound : i will keep like it is : some computer having not sound. i'll open a wiki page for specific chipset if we don't manage to enable them automatically.
- upgrade : the upgrade doesn't run, or at list display nothing (i guess it depends on the framebuffer state at the upgrade moment). However, our way to upgrade (untar) is really bad on usb key (too slow for usb keys). So, probably i could disable upgrades on x86 until we upgrade via squashfs which is instantly and really more powerfull.

so, the real thing to do :
- dosbox : because the .mk is not correct because modifying the files generated by configure and then, requiring to run 2 times make dosbox in some circonstances
- apply changes on the x86_64 builds
and work on the c2 ;-)

i'm interested if somebody is interested to fix scumvm on x86/xu4.
i'm interested if somebody is interested to do dolphin packages (and configgen configuration) ; old versions of dolphin works on x86, so, 2 versions of the package could be done to work on both platform.

**Comment by ron15** [comment]
@nadenislamarre Sorry for the late reply. i've been busy the entire week. I have tried the fix you have given tfor the sound. It worked when I connect the speakers in the audio port. But when using HDMI there's no sound comming out? Is seems this problem is device specific?

**Comment by nadenislamarre** [comment]
2 points.
1) with this sound card, the sound modules are not automatically loaded (i
don't know why).

2) however an other point is that when you've several sound cards, or
several outputs on the same cards,
there is no option in recalbox to change it.

my understanding is that :
echo ""options snd-hda-intel enable=0,1" > /etc/modprobe.d/alsa-intel.conf

is to put the jack output by default because the hdmi output is by default
on this card.
So, removing this file or changing the values (1,0) will probably help you.

Nicolas

**Comment by nadenislamarre** [comment]
What's up on this threads :

- x86 is done.
- x86_64 is done, the pr is here : https://github.com/recalbox/recalbox-buildroot/pull/493
- gamecube and dolphin emulator are done : the pr will be done once the pr on x86_64 is done : https://github.com/nadenislamarre/recalbox-buildroot/commit/d720015bad665d100d128abfaa4159ef8669fd55

What's missing :
- nightly builds (what compile on my pc doesn't always compile on nightly builds)
- autoconfiguration (configgen) for wii/gamecube

**Comment by Akiafroo** [comment]
nice work! will incluide  an option to install it on hard disk like normal distro?

**Comment by nadenislamarre** [comment]
possibly. but not in dual boot or similar things to keep it simple.
however note that wget recalbox.img.gz | gunzip | dd should do the work ;-)

**Comment by Akiafroo** [comment]
thank bro. , if you need help for testing just let me know :)

**Comment by paradadf** [comment]
Finally managed to test it... very nice work Nicolas! Thank you very much!
Some feedback:
* Tried to install the x86_64 (alpha 2) version and my laptop just won't boot from it. It started to create partitions though (It's a 64bit computer). x86 alpha 9 worked without complaints.

* Glupen64 isn't working. Every game I try to launch just brings me back to ES.

* I played Mario Tennis 64 with mupen64plus and Gliden64 (default settings) and the game had slowdowns more or less exactly like on the RPi. Substring opened an issue to update the packages and/or makefiles to add RPi3 where missing. I suppose there should be modifications for the new boards as well.

**Comment by nadenislamarre** [comment]
only 1 n64 gl*64 is built.
c64 doesn't work, not only on x86.
but n64, if confirmed, it's not realted to x86.

I close because the work is finished.

**Comment by nadenislamarre** [comment]
@roba91 thanks. i didn't notice.

**Comment by paradadf** [comment]
@nadenislamarre I wasn't sure if I should open an issue for this or post it here, sorry if I'm mistaken...

Now that the alpha 3 for x86_64 is out I tried again to make it work, but still nothing :(
The "installation" works perfectly with the x86 image on the same USB. Instead of booting from my USB, Windows just starts. My computer has a 64-bit processor, so I don't understand what is happening. Any thoughts?

**Comment by westmc** [comment]
I cannot boot off the x86_64 version either (from a converted VDI on virtualbox or from a USB drive directly).

**Comment by eesprit** [comment]
First things first, a big thank you to @nadenislamarre for your work on the x86* ports, I was waiting for them for ages.

Now I have a question, the x86 version works pretty well installed on an USB pen, but I tried to moved it to my internal SATA SSD, and discover that it failed booting from it. I think that the kernel/initrd is missing the modules that would permit that (as you said earlier in this issue, you explicitely embedded the mass storage drivers in the kernel, not as modules in initrd to have something more simple).

Are there plans to support internal SATA drives ? Might be all in kernel or as modules + initrd.
I would be pleased to help if you are interested in such a feature but don't have the time to do it.
I would also understand if you think that recalbox is more USB/SDcard oriented and so don't want to add more complexity in the boot process.

**Comment by nadenislamarre** [comment]
you're right ; the sata drivers are missing at this boot step.
I should make possible to add drivers into the initramfs build process to
solve it.

i would really like help from you to do it ;-)
i can guide you. just tell me if you feel able to do it.

i don't think that there is a lot of things to do, but i don't know exactly
what must be done :
- in 2 words : modify the initramfs creation script :
package/recalbox-initramfs/recalbox-initramfs.mk (and Config.in) to
- add linux as dependancy to be sure to build the modules before creating
the initramfs
- add modprobe (and probably more) things to compile into busybox (file
package/recalbox-initramfs/busybox.config)
- copying modules (from the .mk file)
- and perhaps something else.

**Comment by nadenislamarre** [comment]
check that secure boot is not enabled in your bios

**Comment by eesprit** [comment]
@nadenislamarre : I'll be happy to help.
For now I am trying to build modules staticly in kernel to be sure that I have correctly identified the ones that are needed.
Then I'll look to do that in the initramfs.
I'll ask if I need any guidelines (I am pretty confident with initrd stuff, but I never looked at the recalbox buildroot process).

Should I work on the rb-4.1.X recalbox branch or should I use any of your forks/branches ?

**Comment by nadenislamarre** [comment]
rb4.1branch is the good one to use.

**Comment by westmc** [comment]
@nadenislamarre 
I'm not running secure boot.  I tried it on another machine as well, but it didn't work.  Both were BIOS and not UEFI.  The 32-bit image worked for me.

**Comment by westmc** [comment]
@nadenislamarre 
I tried compiling using the latest on the rb-4.1.X branch using recalbox-x86_defconfig and I am getting this error...

[ 93%] Linking CXX executable PPSSPPSDL
/home/X/recalbox-buildroot/output/build/ppsspp-v1.2.2/ffmpeg/linux/x86/lib/libavutil.a(dict.o): In function `av_dict_set_int':
/tmp/ppsspp-ffmpeg/libavutil/dict.c:145: multiple definition of `av_dict_set_int'
lib/libCore.a(MediaEngine.cpp.o):MediaEngine.cpp:(.text+0x280): first defined here
collect2: error: ld returned 1 exit status
CMakeFiles/PPSSPPSDL.dir/build.make:816: recipe for target 'PPSSPPSDL' failed
make[4]: *** [PPSSPPSDL] Error 1
CMakeFiles/Makefile2:466: recipe for target 'CMakeFiles/PPSSPPSDL.dir/all' failed
make[3]: *** [CMakeFiles/PPSSPPSDL.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make[2]: *** [all] Error 2
package/pkg-generic.mk:195: recipe for target '/home/X/recalbox-buildroot/output/build/ppsspp-v1.2.2/.stamp_built' failed
make[1]: *** [/home/X/recalbox-buildroot/output/build/ppsspp-v1.2.2/.stamp_built] Error 2
Makefile:36: recipe for target '_all' failed
make: *** [_all] Error 2

**Comment by nadenislamarre** [comment]
thanks to have report that point.
i think i know why it occurs. i will fix that soon.

**Comment by westmc** [comment]
@nadenislamarre 
I tried the x86_64 config as well and got errors involving flags.cpp.  The error is attached.
[make.log.txt](https://github.com/recalbox/recalbox-os/files/554597/make.log.txt)

**Comment by nadenislamarre** [comment]
hi,
thanks for the feedback.
i don't know this error. It looks like a dependancy error.
i'll check that point later.
To be able to finish to compile, just unset dosbox, or directly in .config
or in configs/recalbox_x86_defconfig (require you rerun make x86_defconfig)

**Comment by nadenislamarre** [comment]
ok, i'm just take some minutes to check why it could not work (boot from bios with x86_64 images).
x86_64 computers can boot from bios or efi. i use efi on my computers,
but...

it's possible that i totally forgot to add bios support for the x86_64 boot.
All apologies.
I'll fix, and propose a new build soon.

**Comment by jugir** [comment]
Hello!
Tested the x86 alpha 9 version, looks very promesing.
I tried it with my laptopt, problem is as my laptop resolution is different (hd) as my display (fullhd).
Picture is disorded on the main display and not scaled, it has just has small gameplay window and flashing background surrounding it.

Also my laptop has bot nvidia and intel HD graphics built in.

There should be 

You should add all the known network / wifi adapters as neither my integrated network adapter nor wifi adapter workes, my old realtek based usb stick finally worked and i got this running.

If you need screenshots i can sent.

After this i  corrected will continue testing and try different emulators.

-Jukka

**Comment by jugir** [comment]
Hello,

Here is the problem with display with x86 alpha 9 version,
Also neogeo is not working, i have right bios files.
With raspberry it worked fine.

![file_002](https://cloud.githubusercontent.com/assets/23155520/19840240/649328ea-9efa-11e6-8840-e0b041b276f1.jpeg)
![file_001](https://cloud.githubusercontent.com/assets/23155520/19840241/64ac0d42-9efa-11e6-8d27-afd2c9d12642.jpeg)
![file_000](https://cloud.githubusercontent.com/assets/23155520/19840242/64c3b6ae-9efa-11e6-9ca7-b345a8f92d8f.jpeg)

-Jukka

**Comment by tioneb79** [comment]
I tried to install on my computer as described here https://www.youtube.com/watch?v=ieIQ_xdpRSk but i had a boot failure message. Is the procedure good ? Which files do i need on http://recalbox.remix.free.fr/ ?

thx a lot :)

**Comment by westmc** [comment]
@nadenislamarre I was able to boot off your latest x86_64 image that you provided so your change fixed it for me.  I still can't build from the latest rb-4.1.x.  What am I doing differently from you?

**Comment by nadenislamarre** [comment]
@westmc it depends on your error. note that x86_64 is not compiling on the official server too. i'm on ubuntu 15.04 and the official server is on 16.x, probably it does a difference. (http://build.recalbox.com/)

**Comment by westmc** [comment]
@nadenislamarre I was trying to build with 16.04.  I'll try running with 15.04.  The errors are related to dosbox.h

**Comment by nadenislamarre** [comment]
@westmc can you pastebin me it please ?

**Comment by westmc** [comment]
@nadenislamarre 
[make.log.zip](https://github.com/recalbox/recalbox-os/files/570032/make.log.zip)

**Comment by westmc** [comment]
@nadenislamarre 
I tried again with Ubuntu 15.10 but got essentially the same build error.  I also tried 14.04.5, but no luck.  There must be some other difference between your setup and mine.

[logs.zip](https://github.com/recalbox/recalbox-os/files/570344/logs.zip)

**Comment by nadenislamarre** [comment]
yes. probably i ve some installed packages. i will upgrade on a fresh
ubuntu and fix that

**Comment by nadenislamarre** [comment]
@westmc dosbox compiles on the x86 docker
i will do a test on x86_64 too.
can you send me your config.h where the error occurs ?
note that you can disable dosbox in .config to continue the compilation too.

**Comment by westmc** [comment]
@nadenislamarre 
Which config.h would you like?

I tried Ubuntu 14/15/16 and Debian 8.  All were fresh installs in VirtualBox.  I installed git, build-essentials, openjdk7, mercurial, autoconf, and subversion.  Did a git clone, ran make recalbox-x86_64_defconfig, and then make

**Comment by nadenislamarre** [comment]
From your log :
../../config.h:302:4: error: #error "can't find sizeof(type) of 4 bytes!"


/home/XXX/recalbox-buildroot/output/host/usr/bin/x86_64-buildroot-linux-gnu-g++
-DHAVE_CONFIG_H -I. -I../..  -I../../include -D_LARGEFILE_SOURCE
-D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64
-I/home/XXX/recalbox-buildroot/output/host/usr/x86_64-buildroot-linux-gnu/sysroot/usr/include
-I/home/XXX/recalbox-buildroot/output/host/usr/x86_64-buildroot-linux-gnu/sysroot/usr/include/SDL2
-I/home/XXX/recalbox-buildroot/output/host/usr/x86_64-buildroot-linux-gnu/sysroot/usr/include/SDL
-I/home/XXX/recalbox-buildroot/output/host/usr/x86_64-buildroot-linux-gnu/sysroot/usr/include/SDL2
-D_REENTRANT  -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE
-D_FILE_OFFSET_BITS=64  -O3
-I/home/XXX/recalbox-buildroot/output/host/usr/x86_64-buildroot-linux-gnu/sysroot/usr/include
-I/home/XXX/recalbox-buildroot/output/host/usr/x86_64-buildroot-linux-gnu/sysroot/usr/include/SDL2
-I/home/XXX/recalbox-buildroot/output/host/usr/x86_64-buildroot-linux-gnu/sysroot/usr/include/SDL
-mno-ms-bitfields  -MT core_full.o -MD -MP -MF .deps/core_full.Tpo -c -o
core_full.o core_full.cpp
In file included from ../../include/dosbox.h:23:0,
                 from ../../include/cpu.h:24,
                 from modrm.cpp:19:
../../config.h:282:4: error: #error "sizeof (unsigned char) != 1"
 #  error "sizeof (unsigned char) != 1"
    ^
../../config.h:289:4: error: #error "sizeof (unsigned short) != 2"
 #  error "sizeof (unsigned short) != 2"
    ^
../../config.h:302:4: error: #error "can't find sizeof(type) of 4 bytes!"
 #  error "can't find sizeof(type) of 4 bytes!"
    ^

2016-11-06 6:09 GMT+01:00 westmc <notifications@github.com>:

**Comment by westmc** [comment]
It isn't too bad.  About an hour.  I only gave it 4 cores and 4gb memory so that I could run it in the background.  The files are attached.

[config.zip](https://github.com/recalbox/recalbox-os/files/573993/config.zip)

**Comment by nadenislamarre** [comment]
@westmc it's ok, i reproduce your problem. Let me some time to fix it (not on recalbox the current days)

**Comment by tliron** [comment]
This is terrific (and painstaking) work. :)

I want to suggest a possible reference x86 hardware: the [ECS LIVA Mini PC](http://www.ecs.com.tw/ECSWebSite/Product/Product_LIVA.aspx?DetailID=1560&LanID=0). As of this writing it can be purchased for [US$100](http://www.newegg.com/Product/Product.aspx?Item=N82E16856501007), which includes 32GB storage and 2GB RAM. Like the Raspberry Pi, it sips power via USB and is passively cooled. To put the price in perspective, a complete equivalent Raspberry Pi 3b kit would cost about US$65 (board + box + power supply + heat sinks + SD card).

Things to note about it: it's EFI only and at that requires 64bit operating systems.

I'm excited to try out these builds and report back to you.

What I envision is such a device also running Steam in Big Screen mode, giving gamers access to the very large catalog of games there. (Obviously this hardware isn't powerful enough for very many games, but quit a lot will be fine.) The bottom line is that for US$35 more you can get the same exact features _plus_ the Steam library. When GOG finally gets to releasing a Linux version of their GOG Galaxy client, that can be added, too.

**Comment by tliron** [comment]
And, I can report that the x86_64 works fine on the LIVA. :) This includes sound, a controller (8bitdo), and Ethernet.

I did not get wifi to work, but honestly wifi gives me trouble on standard desktop operating systems on this device.

By the way, people having trouble creating a bootable USB drive: I strongly recommend using [Rufus](https://rufus.akeo.ie/). It's Windows-only (won't work in WINE), but it has never failed me across many different machines, in case where UNetbootin has failed. Though it doesn't find *.img files by default, you can give it the Recalbox *.img and it will do the right thing.

Now to explore how to install it onto the LIVA internal drive ... any tips?

**Comment by substring** [comment]
Hey @tliron

Thank you for your enthusastic feedback :) a word concerning the installation on a hard drive : browse higher up in the thread, you'll have some.hints (init, sata drivers etc ... And we didn't event talk about an installer yet ...)

All the credits go to @nadenislamarre who's been working very hard on porting recalbox to new archs :) You can congratulate him :)

**Comment by tliron** [comment]
I'm working on installing it on a PC's disk and having some trouble with GRUB. Here's my entry:
```
menuentry "Recalbox" {
  insmod part_gpt
  insmod ext2
  insmod gzio
  set root='hd0,gpt4'
  linux /boot/bzImage label=RECALBOX console=tty3 quiet loglevel=0
  initrd /boot/initrd.gz
}
```
Results in a black screen. My question is -- how does the image know how to mount root in a different partition? Is that what `label=` is for? Could there be other ways of specifying it? In my case, I merged the files from the BOOT and RECALBOX partitions into one. How can I tell it to just look for them locally?

**Comment by nadenislamarre** [comment]
you cannot install current images on a hdd. press ctrl alt f3 to get error
messages

**Comment by nadenislamarre** [comment]
i ll add hdd drivers in the boot part and make a script to install. note
that label is used to mount the root partition. merge is not a good idea.
recalbox assume that uou have 3 follow partitions and the middle one is
labelled recalbox. it will be simpler inthe future in some monthes

**Comment by eesprit** [comment]
Yes, I also found out about the partition scheme and was thinking to improve it a little, using several options :
1) if user has given kernel parameters (boot=, recalbox=, share=) than use that
2) if there are labels on partitions (BOOT, RECALBOX, SHARE) than use that
3) fallback to the historical scheme where each partition is at a specific place based on the recalbox partition, so this won't impact anyone

I also found out that when using a NFS share for the whole share (not only a sub-folder) the local share partition is still needed and used (the remote share is mounted over it, which is not really a best practice).

I don't have much free time at the moment, but I'll surely have by the end of the year, so I hope I can spend a little time to propose some changes concerning those points (and also for the kernel/initird thing with disk/SATA drivers).

**Comment by nadenislamarre** [comment]
@eesprit for the moment i would like to keep it simple.

1) i find no advantage to modify these options for the moment
2) the current schema is not the target. The target is only 1 partition for the boot+recalbox with a squashfs partition + a data partition.

**Comment by peachey777** [comment]
Thanks for all your work, this is just awesome. I've tested on 3 systems, and I'm going to test it on another system later tonight. On a cheap little laptop I have it's actually working amazingly well, but a couple powerful machines, it's not. All tests are using x86_64 alpha 4

System 1:
AMD FX-8350
AMD RX 480 8GB
8GB DDR3

System 2:
Intel i7 3770
GTX 1060 6GB
8GB DDR3

Results: Both systems boot fine, however there's no sound, display is stretched instead of widescreen, and the UI/games are very slow and unusable. It seems to me the display drivers for these cards just aren't included in the builds. Is there anyway I could manually install them using a terminal or SSH or something like that?

**Comment by Nicodeme** [comment]
Hello :)

Thanks for all your work and sorry by advance for my poor english.

I have tried your recalbox.remix on an USB key plugged on a NUC (NUC5i3RYH if you want to know) and i have issues. I haven't sound, the trigger in audio panel stay on 0% even if i modify the recalbox.conf in SSH and i can't detect my 8bitdo NES30PRO controller in bluetooth...

I have plugged the NUC on a TV Samsung with HDMI cable and the audio parameter is on HDMI. I don't understand why it doesn't work...

By the way, I have tried to install the img file on a SSD and it don't work. But yours last posts seems that you're working on an installer, can't wait for it !!! :)

Thanks in advance if you have an idea to help me and thanks again for your work !

**Comment by Sp4rtan** [comment]
Hi Nicolas!
Amazing work!
Couldnt sleep and waited for download of 20170104 to finish :D 
First Test on my Lenovo X201 test machine works well  👍 
So how could I get I this beauty on my HDD partition for dualboot? :)

**Comment by nadenislamarre** [comment]
Hi.
I was asked to create a separate issue trackers for everything related to batocera.linux and to my ports.
So, this thread can only continue on https://github.com/nadenislamarre/batocera.linux/issues
